# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

os.getcwd()

'C:\\Users\\filia\\Springboard DS\\Guided Capstone\\1588204778_GuidedCapstoneStep4_and_step5HL.ipynb'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
path = "C:\\Users\\filia\\Springboard DS\\Guided Capstone\\data"
os.chdir(path)
os.getcwd()
df = pd.read_csv('step3_output.csv')
df.head()

Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
0           0       Alyeska Resort   Alaska         3939           2500   
1           1  Eaglecrest Ski Area   Alaska         2600           1540   
2           2     Hilltop Ski Area   Alaska         2090            294   
3           3     Arizona Snowbowl  Arizona        11500           2300   
4           4  Sunrise Park Resort  Arizona        11100           1800   

   trams  fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0      1        0.0          0          2     2  ...             1610.0   
1      0        0.0          0          0     0  ...              640.0   
2      0        0.0          0          0     0  ...               30.0   
3      0        0.0          1          0     2  ...              777.0   
4      0        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 27 columns]

In [3]:
df[df['Name']== 'Big Mountain Resort']

Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
151         151  Big Mountain Resort  Montana         6817           2353   

     trams  fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
151      0        0.0          0          3     2  ...             3000.0   

     Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  \
151           600.0             123.0       72.0            333.0   

     AdultWeekday  AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
151          81.0          81.0              123.0           600.0         1  

[1 rows x 27 columns]

In [4]:
df = df.drop(['Unnamed: 0'], axis=1)
df2 = df #for model2
df3 = df #for model3
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

In [5]:
df.dtypes

Name                  object
state                 object
summit_elev            int64
vertical_drop          int64
trams                  int64
fastEight            float64
fastSixes              int64
fastQuads              int64
quad                   int64
triple                 int64
double                 int64
surface                int64
total_chairs           int64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
dfo = df[['state']] # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
df.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  state_Rhode Island  \
0          0          2     2       0       0  ...                   0   
1          0          0     0       0       4  ...                   0   
2          0          0     0       1       0  ...                   0   
3          1          0     2       2       1  ...                   0   
4          0          1     2       3       1  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
print(y_pred)

[ 58.4942439   92.64658765  78.65440015  75.67686109  70.09287672
  76.86338453  44.33213453  47.42197828  93.56846265  85.59775953
  51.01670484  78.69346265  50.3145564   70.75400953  62.27744703
  93.33994703  41.44883375  73.71690015  63.74033765  40.06260328
 113.4864314   36.65537672  54.18369703  49.2051814   47.53135328
  65.32627515  75.7754939   75.98252515  63.50107984  66.41807203
  35.6973689   58.3692439   81.04136304  56.80186109  75.26670484
  49.2833064   23.26963453  29.03721265  61.3067439   37.0567439
  47.5801814   63.02940015  87.07822828  73.34385328  72.10752515
  58.8379939   73.61729078 166.20127515  37.63291578  60.0958064
  68.37119703  58.95127515  53.5801814   44.66807203  59.48447828
  41.92783765  85.97666578  39.46885328 119.79502515 121.14658765
  55.55283765  52.90244703  42.45908765  81.4551814   64.5567439
  43.2520564   36.42783765  95.83237867  60.72275953  25.9864314
  68.13096265 146.81065015  57.90244703  62.5645564   74.37510328
  70.42783765 

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [12]:
from sklearn.metrics import explained_variance_score, mean_absolute_error
revs = explained_variance_score(y_test, y_pred)
rmae = mean_absolute_error(y_test, y_pred)
print(revs)
print(rmae)

0.9417349856203003
4.859702904398642


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
print(lm.intercept_)

64.17002515385681


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
Coef_states = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(Coef_states.sort_values(by=['Coefficient'], ascending=False))

                       Coefficient
total_chairs          3.041978e+13
fastQuads             1.153221e+13
surface               1.080481e+13
double                9.521603e+12
triple                8.493924e+12
quad                  6.884011e+12
fastSixes             3.418729e+12
trams                 2.937467e+12
state_New York        1.397380e+12
state_Michigan        1.318747e+12
state_Colorado        1.161892e+12
state_California      1.137020e+12
state_Pennsylvania    1.085016e+12
state_New Hampshire   1.000470e+12
state_Wisconsin       1.000470e+12
state_Vermont         9.702422e+11
state_Minnesota       9.388297e+11
state_Utah            9.061092e+11
state_Idaho           8.719336e+11
state_Montana         8.719336e+11
state_Massachusetts   8.361245e+11
state_Washington      7.984620e+11
state_Oregon          7.984620e+11
state_Maine           7.586702e+11
state_New Mexico      7.586702e+11
state_Wyoming         7.163944e+11
state_North Carolina  6.223395e+11
state_Connecticut   

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X2 = df2.drop(['Name','state','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y2 = df2.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler2 = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X2_scaled=scaler2.transform(X2)

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y2 = y2.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y2, test_size=0.25, random_state=1)

In [16]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)
y2_pred = model2.predict(X2_test)
print(y2_pred)

[ 58.92857633  86.97842923  82.97355406  78.2984618   69.14525041
  80.03144655  36.43406899  48.41472639  88.25988904  86.92169314
  52.68924249  80.56916006  52.32863751  65.02377201  64.16355724
  90.19528224  44.28031794  77.15459899  66.17003668  42.5355518
 109.50076805  37.83455065  54.09482977  50.6007551   49.30874883
  65.49128766  68.4155931   75.18176231  62.36079813  65.28085168
  38.49779726  53.7029613   80.99287028  57.0959686   77.31228492
  54.13882066  24.91359701  29.52225226  52.94666328  38.00175817
  47.76993506  62.30555838  90.58516647  68.9697625   70.65939249
  52.12002321  73.21694788 167.69802927  30.81825757  63.01024524
  61.33615404  60.09198454  53.05896635  43.75868436  60.26025052
  42.23533836  86.10181776  32.10114256 121.81565612 116.14895842
  56.98508251  53.64617514  40.77160772  85.57576578  64.37891776
  47.95983305  38.68857699  94.23660442  56.80966989  28.90977654
  64.64378613 151.1202897   55.52175189  64.05261273  75.57758799
  68.487284

In [17]:
revs2 = explained_variance_score(y2_test, y2_pred)
rmae2 = mean_absolute_error(y2_test, y2_pred)
print(revs2)
print(rmae2)

0.9329057841636359
5.240280414532274


In [18]:
print(lm2.intercept_)

64.15173928010691


In [19]:
Coefs = pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])
Coefs.sort_values(by=['Coefficient'], axis=0)

Coefficient
trams                 0.076978
double                0.165526
projectedDaysOpen     0.173338
fastSixes             0.204009
fastEight             0.230663
yearsOpen             0.249735
SkiableTerrain_ac     0.425245
LongestRun_mi         0.427588
total_chairs          0.503686
TerrainParks          0.514425
daysOpenLastYear      0.576510
NightSkiing_ac        0.732038
vertical_drop         0.884491
fastQuads             0.931437
Snow Making_ac        0.974586
quad                  1.030090
averageSnowfall       1.076645
triple                1.312595
surface               1.459882
Runs                  1.952531
summit_elev           4.763727
clusters              6.079200
AdultWeekday         19.860932

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [20]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X3 = df3.drop(['Name','state','summit_elev','AdultWeekend'], axis=1) #'base_elev' was dropped in step3 of capstone

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y3 = df3.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler3 = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X3_scaled=scaler3.transform(X3)

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y3 = y3.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X3_train, X3_test, y3_train, y3_test = train_test_split(X3_scaled, y3, test_size=0.25, random_state=1)

In [21]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)
y3_pred = model3.predict(X3_test)
print(y3_pred)

[ 57.96801743  87.35554402  79.91738751  77.81763589  69.40118182
  78.25577878  37.60425242  48.74311874  88.83169383  88.16057799
  52.19895075  77.81967254  52.68022571  65.94009789  63.83297432
  90.65775825  42.40264541  75.3476444   64.1279911   42.92543599
 109.09765146  38.24857088  55.33770461  50.50575501  49.69921958
  65.66866138  66.4739724   77.17807539  62.96332762  64.76664753
  38.46580737  51.07652617  79.37857725  57.79903787  77.70321776
  53.90578984  25.23656853  29.53446081  54.50317596  39.01171017
  48.98844938  63.01760081  90.51836284  69.16423102  69.01207143
  52.62638052  73.57917532 167.47745472  30.6670313   60.74034735
  62.66908281  59.64463826  53.1047201   45.22889856  60.88577896
  41.94797244  86.62635966  31.73224228 124.32093013 117.1918096
  57.82258539  51.80271487  40.09335063  87.11632124  65.08265523
  48.60089906  37.99971627  98.60692495  57.56273241  25.2098935
  64.56908263 151.73335064  56.49809766  63.6258919   76.3071262
  69.18288863

In [22]:
revs3 = explained_variance_score(y3_test, y3_pred)
rmae3 = mean_absolute_error(y3_test, y3_pred)
print(revs3)
print(rmae3)

0.9338360803618926
5.154053214091007


In [23]:
print(lm3.intercept_)

64.14562716596511


In [24]:
Coefs3 = pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])
Coefs3.sort_values(by=['Coefficient'], axis=0)

Coefficient
fastEight             0.070828
SkiableTerrain_ac     0.153891
double                0.197429
yearsOpen             0.213126
TerrainParks          0.244489
trams                 0.258831
fastSixes             0.305750
LongestRun_mi         0.330602
NightSkiing_ac        0.367087
total_chairs          0.500675
projectedDaysOpen     0.581625
fastQuads             0.661593
averageSnowfall       0.754531
daysOpenLastYear      1.116225
quad                  1.119234
Snow Making_ac        1.122066
surface               1.253443
triple                1.324939
Runs                  1.456798
vertical_drop         1.899505
clusters              2.728355
AdultWeekday         20.068193

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9417349856203003 | 4.859702904398642 |-|
| Model 2. | 0.9329057841636359 | 5.240280414532274 |'state'|
| Model 3. | 0.9338360803618926 | 5.154053214091007 |'state','summit_elev','base_elev'|

Model Selection:

__<font size="5">Model 3</font>__

Drops factors that cannot be changed easily to offset the costs of the new chairlift.

In [25]:
df2.to_csv('for_step_6.csv')
os.listdir()

['coefs&metrics_final_model.csv',
 'for_step_6.csv',
 'step2_output.csv',
 'step3_output.csv']